types/ember-data/index.d.ts

```diff
@@ -119,11 +119,13 @@ export namespace DS {
     ): string;
     const VERSION: string;
 
-    interface AttrOptions<T extends Transform | string | number | boolean | Date = any> {
-        defaultValue?: T extends Transform ? (() => ReturnType<T['deserialize']>) : string | number | boolean | (() => T);
+    interface AttrOptions<T> {
+        defaultValue?: T | (() => T);
         allowNull?: boolean; // TODO: restrict to boolean transform (TS 2.8)
```

Comment:
Semi-related, someone should actually do this. I think the types will not actually allow null for a boolean where `allowNull` is true.